src/dmd/expressionsem.d

```diff
@@ -73,6 +73,45 @@ import dmd.visitor;
 
 enum LOGSEMANTIC = false;
 
+/********************************************************
+ * Perform semantic analysis and CTFE on expressions to produce
+ * a string.
+ * Params:
+ *      buf = append generated string to buffer
+ *      sc = context
+ *      exps = array of Expressions
+ * Returns:
+ *      true on error
```

Comment:
Please return `false` on error

Reply:
I did it this way to be consistent with the practice of 0 meaning success and !=0 being an error code.

Reply:
> If the design is later changed to be an error code

I'd rather have us discuss what is there instead of what could be.
Literally any argument could be made based on "If the design is later changed to [X]".

> Besides, I try to design things so one doesn't have to normally put a ! in front

That seems to be at odds with the principle of avoiding double negative.

Reply:
>I did it this way to be consistent with the practice of 0 meaning success and !=0 being an error code.

Do we use that practice in conjunction with `bool`, or only with `int`? Carrying it over to `bool` seems awfully confusing.

Reply:
I've used it before with things like:
```
C:\cbx>grep "error |=" *.d
expressionsem.d:        error |= checkDeprecated(sc, f);
expressionsem.d:        error |= checkPurity(sc, f);
expressionsem.d:        error |= checkSafety(sc, f);
expressionsem.d:        error |= checkNogc(sc, f);
```
which would be a bit awkward with reversed logic.

dmd frequently returns true to mean error, but not always.
```
declaration.d: * Returns true if error occurs.
declaration.d:     * Returns true if error occurs.
delegatize.d: *      true if error occurs.
expression.d:     * Returns true if error occurs.
expression.d:     * Returns true if error occurs.
expression.d:     * Returns true if error occurs.
expression.d:     * Returns true if error occurs.
expression.d:     * Returns true if error happens.
expression.d:     * Returns true if error occurs.
expressionsem.d: * Returns: `true` if error occur.
func.d:    bool semantic3Errors;               /// true if errors in semantic3 this function's frame ptr
func.d:     * Returns true if error occurs.
globals.d:     * Return true if errors occurred while gagged.
mars.d: *      true if errors in command line
safe.d: *      true if error
statement.d:     *  true if error
statementsem.d:         *      true if error issued
```
> awfully confusing

I don't know why that would be. Whether an int error code is returned or a bool,
```
if (fclose(fp))
    printf("failed\n");
```
is the same, and has been that way forever in C (because C didn't have a bool type until late in the game, and so `int` was used instead).

There's often a `uint errors;` in DMD which gives a count of errors, and that carries over to many `bool error;` declarations.

There are no `bool success;` declarations in DMD.

Reply:
> is the same, and has been that way forever in C

This is not C and D has a bool type.

Reply:
I don't see how: 
```
expressionsem.d:        error |= !checkDeprecated(sc, f);
expressionsem.d:        error |= !checkPurity(sc, f);
expressionsem.d:        error |= !checkSafety(sc, f);
expressionsem.d:        error |= !checkNogc(sc, f);
```

would be awkward.

Reply:
besides if you did
`error = !(checkDeprecated(sc, f) &&  checkPurity(sc, f) && checkSafety(sc, f) && checkNogc(sc, f))`;

the compiler can make use of short-circut-eval and potentially reorder the expressions.

Reply:
This discussion could be made moot by making it an error code from the start, changing the return type to `int` and the DDOC comment to `Returns: non-zero on error`.